mmc-google-analytics/src/test/unit/java/info/novatec/metricscollector/google/RequestBuilderTest.java

```diff
@@ -0,0 +1,188 @@
+        LocalDate startDate = endDate.minusDays(1);
+        requestBuilder.setDateRange(startDate, endDate);
+
+        assertThat(requestBuilder.getDateRange().getStartDate()).isEqualTo(startDate.toString());
+        assertThat(requestBuilder.getDateRange().getEndDate()).isEqualTo(endDate.toString());
+    }
+
+    @Test
+    public void createDimensionFilterClause() {
+        List<DimensionFilter> dimensionFilters =
+            Collections.singletonList(new DimensionFilter().setDimensionName(DataProvider.GA_DIMENSION));
+        DimensionFilterClause filterClause = requestBuilder.createDimensionFilterClause("AND", dimensionFilters);
+
+        assertThat(filterClause.getFilters().size()).isEqualTo(1);
+        assertThat(filterClause.getFilters().get(0).getDimensionName()).isEqualTo(DataProvider.GA_DIMENSION);
+    }
+
+    @Test
+    public void createReport(){
+        String viewId = "123";
+        DimensionFilterClause expectedFilterClause = requestBuilder
+            .addMetrics(Collections.singletonList(DataProvider.GA_METRIC))
+            .addDimensions(Collections.singletonList(DataProvider.GA_DIMENSION))
+            .addDimensionFilter(
+                DataProvider.GA_DIMENSION,
+                DimensionFilterOperators.EXACT,
+                Collections.singletonList(DataProvider.DIMENSION_VALUE))
+            .createDimensionFilterClause("AND", requestBuilder.getDimensionFilters());
+        List<Metric> expectedMetrics = requestBuilder.getMetrics();
+        List<Dimension> expectedDimensions = requestBuilder.getDimensions();
+
+        ReportRequest actualRequest = requestBuilder
+            .buildRequest(viewId)
+            .getReportsRequest()
+            .getReportRequests().get(0);
+
+        assertThat(actualRequest.getMetrics()).isEqualTo(expectedMetrics);
+        assertThat(actualRequest.getDimensions()).isEqualTo(expectedDimensions);
+        assertThat(actualRequest.getDimensionFilterClauses().size()).isEqualTo(1);
+        assertThat(actualRequest.getDimensionFilterClauses().get(0)).isEqualTo(expectedFilterClause);
+        assertThat(actualRequest.getDateRanges().size()).isEqualTo(1);
+        assertThat(actualRequest.getDateRanges().get(0)).isEqualTo(requestBuilder.getDateRange());
+        assertThat(actualRequest.getViewId()).isEqualTo(viewId);
+    }
+
+    @Test
+    public void sendRequest() throws IOException{
+        requestBuilder.buildRequest("123");
+        AnalyticsReporting.Reports reports = mock(AnalyticsReporting.Reports.class);
+        AnalyticsReporting.Reports.BatchGet batchGet = mock(AnalyticsReporting.Reports.BatchGet.class);
+        doReturn(null).when(batchGet).execute();
+        doReturn(batchGet).when(reports).batchGet(requestBuilder.getReportsRequest());
+        doReturn(reports).when(analyticsReporting).reports();
+
+        requestBuilder.sendRequest();
+
+        verify(batchGet, times(1)).execute();
+    }
+
+    @Test(expected = IORuntimeException.class)
```

Comment:
We should also verify that the ex message contains: "Requesting Google Analytics failed!"

Reply:
I could do. But what is the imrovement? Testing the google api? The exception is thrown by AnalyticsReporting, a google class.
If they change the ex message we have to adapt our test

Reply:
Sure, but this message is in our application code. Otherwise, I agree.